<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>Contavect by a-slide</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>Contavect</h1>
        <h2>Analyse DNA contaminants from NGS data</h2>
        <a href="https://github.com/a-slide/ContaVect" class="button"><small>View project on</small>GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h1>
<a name="contavect" class="anchor" href="#contavect"><span class="octicon octicon-link"></span></a>ContaVect</h1>

<h1></h1>

<h2>
<a name="motivation" class="anchor" href="#motivation"><span class="octicon octicon-link"></span></a>Motivation</h2>

<p>Contavect was developped to quantify and caracterize DNA contaminants from gene therapy vector production after NGS sequencing. This automated pipeline can however be used for wider pourpose requiring to identify map NGS datasets consisting of a mix of DNA sequences on multiple references. It combine several features such as reference homologies masking, fastq filtering/adapter trimming, short read alignments, SAM file splitting and generating human readable output.</p>

<h2>
<a name="principle" class="anchor" href="#principle"><span class="octicon octicon-link"></span></a>Principle</h2>

<h3>
<a name="dependencies" class="anchor" href="#dependencies"><span class="octicon octicon-link"></span></a>Dependencies:</h3>

<p>The programm was developed under Linux Mint 16/17 and require a python 2.7 environment.
The following dependencies are required for proper program excecution:</p>

<ul>
<li><a href="https://github.com/lh3/bwa">bwa 0.7.0+</a></li>
<li><a href="https://github.com/samtools/samtools">samtools 0.1.17+</a></li>
<li>blast+ (not required is not no reference masking is to be performed) </li>
</ul><p>In addtion 2 third party python packages are also needed </p>

<ul>
<li><a href="https://github.com/biopython/biopython">Biopython</a></li>
<li><a href="https://github.com/pysam-developers/pysam">pysam</a></li>
</ul><h2>
<a name="get-isfinder" class="anchor" href="#get-isfinder"><span class="octicon octicon-link"></span></a>Get IsFinder</h2>

<ol>
<li>Clone the repository</li>
</ol><div class="highlight highlight-bash"><pre><span class="nv">$ </span>git clone https://github.com/a-slide/ContaVect/ my_folder/
</pre></div>

<ol>
<li>Make the main script excecutable</li>
</ol><div class="highlight highlight-bash"><pre><span class="nv">$ </span>sudo chmod u+x IsFinder.py
</pre></div>

<p>3.Compile the ssw aligner (and add the dynamic library it to the path)</p>

<p>If you wish to perform a step of adapter trimming before mapping you need to complile the dynamic library ssw.so to be able to use the Smith watermann algorithm forked from mengyao's <a href="https://github.com/mengyao/Complete-Striped-Smith-Waterman-Library">Complete-Striped-Smith-Waterman-Library</a></p>

<p>To use the dynamic library libssw.so you may need to modify the LD_LIBRARY_PATH environment
variable to include the library directory (export LD_LIBRARY_PATH=$PWD) or for definitive
inclusion of the lib edit /etc/ld.so.conf and add the path or the directory containing the
library and update the cache by using /sbin/ldconfig as root</p>

<h2>
<a name="usage" class="anchor" href="#usage"><span class="octicon octicon-link"></span></a>Usage</h2>

<p>Prepare the configuration file to include your files and settings as indicated in the template Conf.txt file provided with the source files</p>

<div class="highlight highlight-bash"><pre><span class="nv">$ </span>./ContaVect.py Conf.txt......
</pre></div>

<h2>
<a name="pipeline-development-logbook" class="anchor" href="#pipeline-development-logbook"><span class="octicon octicon-link"></span></a>Pipeline development logbook</h2>

<p>TO BE INCLUDED</p>

<ul>
<li><a href="">Logbook</a></li>
</ul><h2>
<a name="authors-and-contact" class="anchor" href="#authors-and-contact"><span class="octicon octicon-link"></span></a>Authors and Contact</h2>

<p>Adrien Leger - 2014</p>

<ul>
<li><a href="mailto:adrien.leger@gmail.com">adrien.leger@gmail.com</a></li>
<li><a href="mailto:adrien.leger@inserm.fr">adrien.leger@inserm.fr</a></li>
<li><a href="mailto:adrien.leger@univ-nantes.fr">adrien.leger@univ-nantes.fr</a></li>
<li><a href="https://github.com/a-slide">Github</a></li>
<li><a href="http://www.atlantic-gene-therapies.fr/">Atlantic Gene Therapies - INSERM 1089</a></li>
</ul>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/a-slide/ContaVect/zipball/master" class="button">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/a-slide/ContaVect/tarball/master" class="button">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/a-slide/ContaVect"></a> is maintained by <a href="https://github.com/a-slide">a-slide</a>.</p>

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the Architect theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

  
  </body>
</html>